action/protocol/staking/handlers.go

```diff
@@ -691,20 +691,21 @@ func (p *Protocol) handleCandidateRegister(ctx context.Context, act *action.Cand
 	// cannot collide with existing identifier
 	candID := owner
 	if !featureCtx.CandidateIdentifiedByOwner {
-		id, err := p.generateCandidateID(owner, blkCtx.BlockHeight, csm)
-		if err != nil {
+		// cannot collide with existing identifier
+		if csm.GetByIdentifier(owner) != nil {
 			return log, nil, &handleError{
-				err:           errors.Wrap(err, "failed to generate candidate ID"),
+				err:           ErrInvalidOwner,
 				failureStatus: iotextypes.ReceiptStatus_ErrCandidateAlreadyExist,
 			}
 		}
-		candID = id
-		if csm.GetByIdentifier(candID) != nil {
+		id, err := p.generateCandidateID(owner, blkCtx.BlockHeight, csm)
```

Comment:
by moving the conflict check ahead, `generateCandidateID` will always return the original owner address, the for loop will never enter. this would cause 1 address can register only once

as discussed last evening, should remove the 3 lines, and let for loop to generate a random identifier. Then, the same address (after transfer ownership) can register again

Reply:
Yes, the first point is correct that only owner address will be used as identifier.
You might miss some part of the discussion. After the hard fork, preventing the same address to generate a new candidate is on purpose, to prevent confusion: CandidateByOwner and CandidateByIdentifier returns two different candidates given the same address. The reason I didn't ask @envestcc to delete the logic in `generateCandidateID` in this PR is that it really doesn't matter to keep it.